API_DESIGN_GUIDE.md

```diff
@@ -0,0 +1,86 @@
+# API Design Guide
+
+Before reading this document, please read the
+[Keras API design guidelines](https://github.com/keras-team/governance/blob/master/keras_api_design_guidelines.md).
+
+Below are some design considerations specific to KerasNLP.
+
+## Philosophy
+
+- **Let user needs be our compass.** Any modular building block that NLP
+  practitioners need is in scope, whether it's data loading, augmentation, model
+  building, evaluation metrics, or visualization utils.
+
+- **Be resolutely high-level.** Even if something is easy to do by hand in 5
+  lines, package it as a one liner.
+
+- **Balance ease of use and flexibility.** Simple things should be easy, and
+  arbitrarily advanced use cases should be possible. There should always be a
+  "we need to go deeper" path available to our most expert users.
+
+- **Grow as a platform and as a community.** KerasNLP development should be
+  driven by the community, with feature and release planning happening in
+  the open on GitHub.
+
+## Avoid new dependencies
+
+The core dependencies of KerasNLP are Keras, NumPy, TensorFlow, and
+[Tensorflow Text](https://www.tensorflow.org/text).
+
+We strive to keep KerasNLP as self-contained as possible, and avoid adding
+dependencies to projects (for example NLTK or spaCy) for text preprocessing.
+
+In rare cases, particularly with tokenizers and metrics, we may need to add
+an external dependency for compatibility with the "canonical" implementation
+of a certain technique. In these cases, avoid adding a new package dependency,
+and add installation instructions for the specific symbol:
+
+```python
+try:
+    import rouge_score
+except ImportError:
+    pass
+
+class RougeL(keras.metrics.Metric):
+    def __init__(self):
+        if rouge_score is None:
+            raise ImportError(
+                'RougeL metrics requires the rouge_score package. '
+                '`pip install rouge-score`.')
+```
+
+## Keep computation inside TensorFlow graph
```

Comment:
Maybe worth pointing out the difference between "TF graph" (`tf.function`) and "XLA compilable" (`tf.function(jit_compile=True)`)